flocker/node/functional/test_docker.py

```diff
@@ -208,6 +208,51 @@ def started(_):
         d.addCallback(started)
         return d
 
+    def test_list_image_data_cached(self):
+        """
+        ``DockerClient.list`` will only an inspect an image ID once, caching
+        the resulting data.
+        """
+        image_name = u"openshift/busybox-http-app:latest"
+        name = random_name(self)
+        d = self.start_container(name, image_name=image_name)
+
+        def started(client):
+            listing = client.list()
+
+            def listed(_):
+                docker = Client()
+                data = docker.inspect_container(self.namespacing_prefix + name)
+                # This is kind of nasty, but NamespacedDockerClient represents
+                # its client via a proxying attribute, so we can't assume
+                # client here directly has an _image_cache.
+                if isinstance(client, NamespacedDockerClient):
+                    image_cache = client._client._image_cache
+                else:
+                    image_cache = client._image_cache
+                self.assertIn(data['Image'], image_cache)
```

Comment:
We don't really care about the implementation details of the cache.  Ideally, the only difference the cache makes is that some things get faster.  So, ideally, the best way to justify this change is to demonstrate that the desired outputs are unchanged (ie, it still "works" - hopefully just by passing all of the existing tests) and to demonstrate that it provides those outputs faster than it did before.

Demonstrating things about fastness requires benchmarks.  We've skipped over that step for the past two weeks because of Reasons.  Since we're still lacking any general tool chain or infrastructure for having benchmarks, it's probably okay to skip over it for this change as well.

But what we don't want is new unit "tests" that just reveal optimization-motivated implementation details.  The thing we want to prove is that this is *fast*.  We don't care if there's an `_image_cache`.

Itamar's suggestion for a test, involving patching the Docker client's `inspect_image` method is a sort of compromise between these two.  It drops the assertions this test currently makes about the irrelevant `_image_cache` - that's good.  In place of measuring how fast the old and new code is, though, it supposes that fewer calls to `inspect_image` must mean things are getting faster - so it proves that `inspect_image` is not called in a certain case (where we know it was called before).  It's still not ideal but it is a little better.